docs/geps/gep-01.rst

```diff
@@ -121,7 +120,7 @@ or value change has a reference to a law or a source.
             de: Regelsatz
             en: Standard rate
 
-2. The `description` key has two sub-keys `de` and `en`, which
```

Comment:
I just got stuck on this example; I think it is a great case in point to illustrate why an approach with dicts etc. will be much clearer.
- In `arbeitsl_geld_2.yaml`, the parameters `a2an1` ... `a2eg3` essentially repeat the description six times. This is very easy to get wrong at some point / forget a copy & paste.
- I find it very hard to see the schedule for a given point in time. You essentially have to check six (seven with `a2eg3ki`... who comes up with this kind of policies?) over 90 lines.

I suggest we walk through this example to see what we might want to do in general. My proposal (certainly not complete), explanations below:

```yaml
e_anr_frei:
    name:
      de: Anrechnungsfreie Einkommensanteile
      en: Income shares not subject to transfer withdrawal
  description:
    de: Einkommensanteile, die anrechnungsfrei bleiben. § 30 SGB II. Seit 01.10.2005 zudem definiert durch Freibetrag in § 11 SGB II, siehe auch § 67 SGB II. Seit 01.04.2011 § 11b (2) SGB II (neugefasst durch B. v. 13.05.2011 BGBl. I S. 850. Artikel 2 G. v. 24.03.2011 BGBl. I S. 453).
    en: ??? Do we repeat everything or just translation of the first bit?
  type: piecewise_linear
  value:
    2005-01-01:
      value: 
        400: 0.15
        900: 0.3
        1500: 0.15
      note: G. v. 24.12.2003 BGBl. I S. 2954.
    2005-10-01:
      value: 
        100: 1
        800: 0.2
        1200: 0.1
      note: G. v. 14.08.2005 BGBl. I S. 2407.
    2011-04-01:
      value: 
        100: 1
        1000: 0.2
        1200: 0.1
      note: SGB II. B. v. 13.5.2011 BGBl. I S. 850. Artikel 5 G. v. 20.12.2011 BGBl. I 2854. G. v. 14.08.2005 BGBl. I S. 2407.
```
Notes:
* `e_...` to be consistent with `e_st`
* I do not like the value(s).value nested construct that we have, but could not come up with anything better for now
* Stuff like `piecewise_linear` will need a clear definition somewhere
* Not sure how we want to handle the cases where only one parameter of six changes.
* Neither for the current `a2eg3ki` -- I would tend to repeat the whole thing and make clear the former is for households without, the second with kids (I would do that in the code for sure, not so obvious in the YAML).

Reply:
Maybe we could use `upper_thresholds_rates` instead of (the inner) `value`? This would mean that the keys of `value[date]` would be differ across parameters, but it would become very descriptive.

Reply:
I see the benefits of this approach when the param file shall be easier to read. An even stronger case can be made for the Wohngeld parameters by household size (see the changes in #134). One has to be careful though to limit this to policies which are, structurally, stable over time. This won't hence work for the income tax tariff.

Is this how the param in the code would look like: `params["e_anr_frei"][100]`? I find this borderline acceptable. We certainly need to balance readability of the param file and readability of the code. After all, the param file is primarily a data source and only secondarily a documentation. A more user-friendly documentation could be compiled from this fairly easy, e.g. extending #122, while not touching the way these params are used in the code.

> * Stuff like `piecewise_linear` will need a clear definition somewhere

These are, after all, only the numeric parameters. In my view, the docstring of the high-level function (the one called from `tax_transfer()` is the natural place to describe what a function is actually doing.

Reply:
> I see the benefits of this approach for the purpose of making the param file easier to read. An even stronger case can be made for the Wohngeld parameters by household size (see the changes in #134). One has to be careful though to limit this to policies which are, structurally, stable over time. This won't hence work for the income tax tariff.

If this latter case is prevalent, it would be a good reason to put the "type" parameter inside the value.date objects (along with `value` [`upper_thresholds_rates`] / `note`). We have a misunderstanding about what I have in mind there, see below.

> Is this how the param in the code would look like: `params["e_anr_frei"][100]`? I find this borderline acceptable. 

No. See below.

> We certainly need to balance readability of the param file and readability of the code. After all, the param file is primarily a data source and only secondarily a documentation. A more user-friendly documentation could be compiled from this fairly easy, e.g. extending #122, while not touching the way these params are used in the code.

> > Stuff like `piecewise_linear` will need a clear definition somewhere
> 
> These are, after all, only the numeric parameters. In my view, the docstring of the high-level function (the one called from `tax_transfer()` is the natural place to describe what a function is actually doing.

We will have a set of generic functions doing this, `piecewise_linear` will be one of them. The `piecewise_polynomial` we [discussed about before](https://gettsim.zulipchat.com/#narrow/stream/212222-general/topic/GEPs/near/185481518) will be another one. `step_function` another one. Names of course not pinned down. We could have some with jumps etc.. The point is there is nothing special about the precise application in the context of social assistance / wohngeld / tax schedules / ... and we do not need to re-write the same things over and over.

Reply:
> If this latter case is prevalent, it would be a good reason to put the "type" parameter inside the value.date objects (along with `value` [`upper_thresholds_rates`] / `note`). We have a misunderstanding about what I have in mind there, see below.

Now I see where this is going. I don't think the latter is the prevalent case, but there will be a bunch of generic cases which can hardly be generalized. Vorsorgeaufwendungen and income test for Wohngeld come to my mind.

Reply:
My understanding is that the functions in question are usually rather generic. Some of the super-old tax schedules were actually so ingenious that they had jumps at the cutoffs, so we want to keep that case in mind. Other than that, all these schedules---whether they are taxes or transfer-reducing---seem pretty straightforward.

The hard part, and indeed the one you cannot generalise much, is coming up with the quantities. E.g., once you have taxable income the schedule is some lines of generic code. Coming up with that number requires several hundred or thousand lines of code, though.

Is that any different with the examples you were citing?

Reply:
OK, I guess one could count vorsorgeaufwendungen as quantities. The wohngeld income test is basically a share of income that is subtracted. But this share is endogenous to whether or not one pays income taxes, health care contributions or old-age pension contributions. So this might be done with a general linear function, with the "slope" not directly coming from the parameters.